src/cache.rs

```diff
@@ -1186,9 +1190,41 @@ fn with_parent(path: &OsStr, parent: Option<PathBuf>) -> PathBuf {
 
 /// Normalize the path of a file for unique identification in the cache.
 ///
-/// If an IO error occurs here, `None` is returned.
-pub fn normalize_path(path: &Path) -> io::Result<OsString> {
-    path.canonicalize().map(|p_| p_.as_os_str().to_os_string())
+/// This implementation (including the commend below) was taken from cargo-util.
+///
+/// CAUTION: This does not resolve symlinks (unlike
+/// [`std::fs::canonicalize`]). This may cause incorrect or surprising
+/// behavior at times. This should be used carefully. Unfortunately,
+/// [`std::fs::canonicalize`] can be hard to use correctly, since it can often
+/// fail, or on Windows returns annoying device paths. This is a problem Cargo
+/// needs to improve on.
```

Comment:
Do you know what are the "incorrect and surprising" behavior at times? In particular, if it's just that it fails to normalize two paths to the same thing although they are equals, that's fine (we're just going to waste a bit of cache by duplicating the file). However, if the behavior is that it can equate two paths that aren't actually the same, that's a whole different story (with correctness and security implications)

Reply:
I think the surprises are mostly to do with symlinks. For example, this implementation will normalize `a/../c` to `c`,
but if `a` is a symlink to `b/a` then other implementations (e.g. `std::fs::canonicalize`) will normalize it to `b/c` because they resolve the symlinks before the `..`.

I looked a bit more into what rust-analyzer does, and they've decided that correct canonicalization is just [too hard](https://github.com/rust-lang/rust-analyzer/pull/14430#pullrequestreview-1411530295), and so they use absolute-but-not-canonical paths and rely on the OS to figure out what files they resolve to.

Reply:
Ok, so it sounds like indeed cargo's normalization tries to be less smart and might equate less things (which would make sense, and is fine). Thanks for the explanation :+1: